File: 백준/Gold/6497. 전력난/전력난.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

const int MX =  200020;
int m,n;
int x,y;
long long z;

int par[MX];
vector<pair<long, pair<int,int>>> vec;

int find(int x){
    if(par[x] ==x) return x;
    else return par[x] = find(par[x]);
}

void union_(int a, int b){
    a = find(a);
    b = find(b);
    
    if(a==b) return;
    
    par[a] = b;
}

void solution(){
    vec = {};
    for(int i = 0; i < m; i++){
        par[i] = i;
    }
    
    for(int i = 0; i < n; i++){
        cin >> x >> y >> z;
        
        vec.push_back({z,{x,y}});
    }
    
    sort(vec.begin(),vec.end());

    long sum = 0;
    long used = 0;
    for(int i = 0; i < n; i++){
        int a = find(vec[i].second.first);
        int b = find(vec[i].second.second);
        long c = vec[i].first;
        
        sum += c;
        if(a == b) continue;
      
        used += c;
        union_(a,b);
    }
    
    cout << (sum - used) << '\n';
}

int main()
{
    int t;
    while(true){
        cin >> m >> n;
        if(m == 0 && n == 0) break;
        solution();
    }
    
    return 0;
}

File: 백준/Gold/6497. 전력난/전력난.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.StringTokenizer;

public class Main {
	static final int MX = 200020;
	static int par[] = new int[MX];
	static ArrayList<long[]> vec = new ArrayList<>();
	static StringBuilder sb = new StringBuilder();
	static int V, E;
	static BufferedReader br;
	static StringTokenizer st;

	public static void main(String[] args) throws Exception {
		br = new BufferedReader(new InputStreamReader(System.in));
		while (true) {
			st = new StringTokenizer(br.readLine());
			V = Integer.parseInt(st.nextToken());
			E = Integer.parseInt(st.nextToken());
			if (V == 0 && E == 0)
				break;
			solution();
		}

		System.out.println(sb);
	}

	static void solution() throws Exception {
		init();
		vec.clear();
		long ret = 0;
		for (int i = 0; i < E; i++) {
			st = new StringTokenizer(br.readLine());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			long c = Long.parseLong(st.nextToken());

			vec.add(new long[] { c, a, b });
			ret += c;
		}

		Collections.sort(vec, (o1, o2) -> Long.compare(o1[0], o2[0]));

		for (long[] cur : vec) {
			long c = cur[0];
			int a = (int) cur[1];
			int b = (int) cur[2];
			if (find(a) == find(b))
				continue;
			union(a, b);
			ret -= c;
		}

		sb.append(ret).append("\n");
	}

	static void init() {
		for (int i = 0; i < MX; i++) {
			par[i] = i;
		}
	}

	static int find(int a) {
		if (par[a] == a)
			return a;
		return par[a] = find(par[a]);
	}

	static void union(int a, int b) {
		a = find(a);
		b = find(b);
		if (a == b)
			return;
		par[b] = a;
	}
}